python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,780 @@
+
+        :param dict data: The data passed down by the client. At a minimum,
+            the dict should contain the following keys: project_id, entity_id,
+            entity_type, pipeline_configs, and user.
+        """
+        # If we weren't sent a usable entity id, we can just query the first one
+        # from the project. This isn't a big deal for us, because we're only
+        # concerned with picking the correct environment when we bootstrap
+        # during a caching operation. This is going to be the situation when
+        # a page is pre-loading entity-level commands on page load, as opposed to
+        # passing down a specific entity that's been selected in an already-loaded
+        # page.
+        if "entity_id" in data and data["entity_id"] == -1:
+            if data["entity_type"] == "Project":
+                data["entity_id"] = data["project_id"]
+            else:
+                temp_entity = self._engine.shotgun.find_one(
+                    data["entity_type"],
+                    [["project", "is", dict(type="Project", id=data["project_id"])]],
+                )
+                data["entity_id"] = temp_entity["id"]
+
+        project_entity, entities = self._get_entities_from_payload(data)
+        entity = entities[0]
+
+        # We can end up getting requests for actions for any entity type
+        # that exists in Shotgun. Many of those we don't want to provide
+        # commands for, so we can just notify the client that it's an
+        # unsupported entity type.
+        env_name = self.__pick_environment(project_entity, entity)
+
+        if env_name is None:
+            self.host.reply(dict(retcode=self.UNSUPPORTED_ENTITY_TYPE))
+            return
+
+        manager = sgtk.bootstrap.ToolkitManager()
+        manager.base_configuration = self.BASE_CONFIG_URI
+        pcs = self._get_pipeline_configurations(
+            manager,
+            project_entity,
+        ) or [dict(id=None, name="Primary")]
+
+        # We'll need to pass up the config names in order along with a dict of
+        # pc_name => commands.
+        pc_names = [pc["name"] for pc in pcs]
+        all_actions = dict()
+        config_data = dict()
+
+        for pc in pcs:
+            # The hash that acts as the key we'll use to look up our cached
+            # data will be based on the entity type and the pipeline config's
+            # descriptor uri. We can get the descriptor from the toolkit
+            # manager and pass that through along with the entity type from SG
+            # to the core hook that computes the hash.
+            manager.pipeline_configuration = pc["id"]
+            pc_descriptor = manager.resolve_descriptor(project_entity)
+            pc_key = pc_descriptor.get_uri()
+
+            pc_data = dict()
+            pc_data["contents_hash"] = self._get_contents_hash(
+                pc_descriptor,
+                self._get_site_state_data(),
+            )
```

Comment:
debug logging for what is being returned from all hooks is helpful for troubleshooting i think.